In [11]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [13]:
# Create path
path =  r'C:\Users\Marker\Videos\Instakart Basket Analysis 01-25'

In [23]:

# Import df_ords_prods_merge dataframe/ data
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_refined.pkl'))


In [87]:
# Checking the Shape of the Dataframe
df_ords_prods_merge.shape

(32404859, 24)

In [24]:
# Step 2
# Create subset
df_ords_prods_merge_sub = df_ords_prods_merge[:1000000]

In [27]:
df_ords_prods_merge_sub.groupby(['department_id']).agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.577493
2                17.320781
3                16.084944
4                17.530458
5                14.763075
6                16.658449
7                17.031590
8                15.076662
9                15.447580
10               18.681852
11               15.447411
12               14.327957
13               16.548642
14               16.960241
15               16.121948
16               17.803851
17               15.593633
18               19.674252
19               16.899756
20               16.255442
21               25.535479

In [91]:
# Grouping by 'department_id' and 'order_number', and calculating the mean using agg() function on entire dataframe
df_ords_prods_merge.groupby(['department_id']).agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [33]:

df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   3139998      138            28          6                 11   
1   1977647      138            30          6                 17   
2    389851      709             2          0                 21   
3    652770      764             1          3                 13   
4   1813452      764             3          4                 17   

   days_since_prior_order  add_to_cart_order  reordered _merge  \
0                     3.0                  5          0   both   
1                    20.0                  1          1   both   
2                     6.0                 20          0   both   
3                     7.0                 10          0   both   
4                     9.0                 11          1   both   

     price_range_loc     Busiest_day    Busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regularly busy           Most orders  
1  Mid-range product  Regularly busy  Regularly busy        Average orders  
2  Mid-range product     Busiest day    Busiest days        Average orders  
3  Mid-range product  Regularly busy    Slowest days           Most orders  
4  Mid-range product      Least busy    Slowest days        Average orders

In [85]:
# Step 3

Comparing the subset and entire dataframe: 

The entire dataframe generally has higher mean order numbers across departments, indicating it captures a broader range of customer behaviours than the subset. 
Notable Differences: 

Department 21 has the highest mean in both, but it’s higher in the subset (25.53) than the entire dataset (22.90). Department 10 greatly increases the dataframe as a whole (20.19 vs. 18.68). 

Consistent Departments: 
Some departments, like 13 and 17, show little difference between the two datasets, while others, like 3, 9, and 12, have notable variations.

Overall, the entire dataframe reveals more frequent ordering patterns across departments compared to the subset.

In [37]:
# Step 4
# Creating new column "max_order"
df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\Marker\AppData\Local\Temp\ipykernel_596\93056017.py:3: FutureWarning: The provided callable <function max at 0x0000020BDD2AB100> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)


In [39]:
# Checking for new column
df_ords_prods_merge.head(20)

product_id                product_name  aisle_id  department_id  prices  \
0            1  Chocolate Sandwich Cookies        61             19     5.8   
1            1  Chocolate Sandwich Cookies        61             19     5.8   
2            1  Chocolate Sandwich Cookies        61             19     5.8   
3            1  Chocolate Sandwich Cookies        61             19     5.8   
4            1  Chocolate Sandwich Cookies        61             19     5.8   
5            1  Chocolate Sandwich Cookies        61             19     5.8   
6            1  Chocolate Sandwich Cookies        61             19     5.8   
7            1  Chocolate Sandwich Cookies        61             19     5.8   
8            1  Chocolate Sandwich Cookies        61             19     5.8   
9            1  Chocolate Sandwich Cookies        61             19     5.8   
10           1  Chocolate Sandwich Cookies        61             19     5.8   
11           1  Chocolate Sandwich Cookies        61             19     5.8   
12           1  Chocolate Sandwich Cookies        61             19     5.8   
13           1  Chocolate Sandwich Cookies        61             19     5.8   
14           1  Chocolate Sandwich Cookies        61             19     5.8   
15           1  Chocolate Sandwich Cookies        61             19     5.8   
16           1  Chocolate Sandwich Cookies        61             19     5.8   
17           1  Chocolate Sandwich Cookies        61             19     5.8   
18           1  Chocolate Sandwich Cookies        61             19     5.8   
19           1  Chocolate Sandwich Cookies        61             19     5.8   

    order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    3139998      138            28          6                 11   
1    1977647      138            30          6                 17   
2     389851      709             2          0                 21   
3     652770      764             1          3                 13   
4    1813452      764             3          4                 17   
5    1701441      777            16          1                  7   
6    1871483      825             3          2                 14   
7    1290456      910            12          3                 10   
8     369558     1052            10          1                 20   
9     589712     1052            15          1                 12   
10   1798115     1379             1          5                 11   
11    497193     1480            36          3                 13   
12   1727826     1480            38          4                 13   
13   2694698     1480            39          2                 12   
14   3171753     1494             6          3                 10   
15    186761     1494             7          2                  8   
16    317121     1494             9          5                  8   
17    200390     1540             4          2                  0   
18    374436     1540             5          1                 21   
19   2920553     1540             6          1                  0   

    days_since_prior_order  add_to_cart_order  reordered _merge  \
0                      3.0                  5          0   both   
1                     20.0                  1          1   both   
2                      6.0                 20          0   both   
3                      7.0                 10          0   both   
4                      9.0                 11          1   both   
5                     26.0                  7          0   both   
6                     30.0                  2          0   both   
7                     30.0                  1          0   both   
8                     19.0                  1          0   both   
9                     15.0                  2          1   both   
10                     7.0                  1          0   both   
11                     7.0                  1          0   both   
12                     

In [47]:
# Assign loyalty flag
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal Customer'

df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular Customer'

df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New Customer'

In [49]:
# Check loyalty flags
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular Customer    15876776
Loyal Customer      10284093
New Customer         6243990
Name: count, dtype: int64

In [51]:
# Check Loyalty flag in dataframe
df_ords_prods_merge[['user_id', 'loyalty_flag', 'order_number','max_order']].head(20)

user_id      loyalty_flag  order_number  max_order
0       138  Regular Customer            28         32
1       138  Regular Customer            30         32
2       709      New Customer             2          5
3       764      New Customer             1          3
4       764      New Customer             3          3
5       777  Regular Customer            16         26
6       825      New Customer             3          9
7       910  Regular Customer            12         12
8      1052  Regular Customer            10         20
9      1052  Regular Customer            15         20
10     1379      New Customer             1         10
11     1480    Loyal Customer            36         43
12     1480    Loyal Customer            38         43
13     1480    Loyal Customer            39         43
14     1494  Regular Customer             6         13
15     1494  Regular Customer             7         13
16     1494  Regular Customer             9         13
17     1540  Regular Customer             4         28
18     1540  Regular Customer             5         28
19     1540  Regular Customer             6         28

In [57]:
# Step 5
# Min, max and mean of each loyalty flag
df_ords_prods_merge.groupby(['loyalty_flag']).agg({'prices':['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal Customer    10.386336  1.0  99999.0
New Customer      13.294670  1.0  99999.0
Regular Customer  12.495717  1.0  99999.0

Loyal customers tend to purchase lower-cost items, with an average price of 10.39. Regular customers fall in the middle, averaging 
12.5 and the new customers pay the most.The extremely high maximum cost of 99999 suggests possible outliers or data errors.

In [61]:
# Step 6
# Creating column for average spender
df_ords_prods_merge['average_spend']= df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\Marker\AppData\Local\Temp\ipykernel_596\2834949931.py:3: FutureWarning: The provided callable <function mean at 0x0000020BDD2ABB00> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  df_ords_prods_merge['average_spend']= df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)


In [67]:
# Creating flag for low and high spender
df_ords_prods_merge.loc[df_ords_prods_merge['average_spend']<10,'spender_flag']='Low Spender'

df_ords_prods_merge.loc[df_ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High Spender'

In [69]:
# Check spender flags
df_ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low Spender     31770614
High Spender      634245
Name: count, dtype: int64

In [71]:
pd.options.display.max_rows = 100

In [73]:
df_ords_prods_merge[['user_id', 'average_spend', 'spender_flag', 'prices']].head(100)

user_id  average_spend  spender_flag  prices
0       138       6.935811   Low Spender     5.8
1       138       6.935811   Low Spender     5.8
2       709       7.930208   Low Spender     5.8
3       764       4.972414   Low Spender     5.8
4       764       4.972414   Low Spender     5.8
5       777       6.935398   Low Spender     5.8
6       825       5.957576   Low Spender     5.8
7       910       6.680000   Low Spender     5.8
8      1052       7.162500   Low Spender     5.8
9      1052       7.162500   Low Spender     5.8
10     1379       5.813636   Low Spender     5.8
11     1480       5.081633   Low Spender     5.8
12     1480       5.081633   Low Spender     5.8
13     1480       5.081633   Low Spender     5.8
14     1494       6.803797   Low Spender     5.8
15     1494       6.803797   Low Spender     5.8
16     1494       6.803797   Low Spender     5.8
17     1540       6.607122   Low Spender     5.8
18     1540       6.607122   Low Spender     5.8
19     1540       6.607122   Low Spender     5.8
20     1540       6.607122   Low Spender     5.8
21     1540       6.607122   Low Spender     5.8
22     1540       6.607122   Low Spender     5.8
23     1540       6.607122   Low Spender     5.8
24     1540       6.607122   Low Spender     5.8
25     1540       6.607122   Low Spender     5.8
26     1540       6.607122   Low Spender     5.8
27     1540       6.607122   Low Spender     5.8
28     1540       6.607122   Low Spender     5.8
29     1540       6.607122   Low Spender     5.8
30     1540       6.607122   Low Spender     5.8
31     1540       6.607122   Low Spender     5.8
32     1540       6.607122   Low Spender     5.8
33     1540       6.607122   Low Spender     5.8
34     1598       7.123529   Low Spender     5.8
35     1598       7.123529   Low Spender     5.8
36     1598       7.123529   Low Spender     5.8
37     1647       4.853521   Low Spender     5.8
38     1647       4.853521   Low Spender     5.8
39     1647       4.853521   Low Spender     5.8
40     1647       4.853521   Low Spender     5.8
41     1647       4.853521   Low Spender     5.8
42     2833       7.377143   Low Spender     5.8
43     2850       8.054153   Low Spender     5.8
44     2857       8.353793   Low Spender     5.8
45     3010       6.452800   Low Spender     5.8
46     3029       7.960606   Low Spender     5.8
47     3395       6.322222   Low Spender     5.8
48     3904       7.511111   Low Spender     5.8
49     4006       7.441100   Low Spender     5.8
50     4006       7.441100   Low Spender     5.8
51     4006       7.441100   Low Spender     5.8
52     4006       7.441100   Low Spender     5.8
53     4006       7.441100   Low Spender     5.8
54     4122       7.598765   Low Spender     5.8
55     4141       6.033333   Low Spender     5.8
56     4343       7.393939   Low Spender     5.8
57     4553       5.715789   Low Spender     5.8
58     4845       7.229630   Low Spender     5.8
59     6866       4.807143   Low Spender     5.8
60     7275       4.620833   Low Spender     5.8
61     7669       9.106757   Low Spender     5.8
62     7669       9.106757   Low Spender     5.8
63     7669       9.106757   Low Spender     5.8
64     7981       7.115714   Low Spender     5.8
65     7981       7.115714   Low Spender     5.8
66     7981       7.115714   Low Spender     5.8
67     7981       7.115714   Low Spender     5.8
68     7981       7.115714   Low Spender     5.8
69     8703       6.593459   Low Spender     5.8
70     8723       5.677397   Low Spender     5.8
71     8735       5.850000   Low Spender     5.8
72     9293      11.857143  High Spender     5.8
73     9293      11.857143  High Spender     5.8
74     9467       7.709091   Low Spender     5.8
75     9694       7.075354   Low Spender     5.8
76     9694       7.075354   Low Spender     5.8
77     9694       7.075354   Low Spender     5.8
78     9694       7.075354   Low Spender     5.8
79     9694       7.075354   Low Spender     5.8
80     9694       7.075354   Low Sp

In [75]:
# Step 7
# Create a column for frequent users
df_ords_prods_merge['Customer_frequency'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\Marker\AppData\Local\Temp\ipykernel_596\3113837384.py:3: FutureWarning: The provided callable <function median at 0x0000020BDD3EDF80> is currently using SeriesGroupBy.median. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "median" instead.
  df_ords_prods_merge['Customer_frequency'] = df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [79]:
# Creating a flag on order frequency
df_ords_prods_merge.loc[df_ords_prods_merge['Customer_frequency'] > 20, 'frequency_flag'] = 'Nonfrequent Customer'

df_ords_prods_merge.loc[(df_ords_prods_merge['Customer_frequency']>10) & (df_ords_prods_merge['Customer_frequency']<=20),'frequency_flag'] = 'Regular Customer'

df_ords_prods_merge.loc[df_ords_prods_merge['Customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [81]:
# Check frequency of flags

df_ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent Customer       22790523
Regular Customer         6927608
Nonfrequent Customer     2686728
Name: count, dtype: int64

In [83]:
df_ords_prods_merge[['user_id','days_since_prior_order', 'Customer_frequency', 'frequency_flag']].head(100)

user_id  days_since_prior_order  Customer_frequency        frequency_flag
0       138                     3.0                 8.0     Frequent Customer
1       138                    20.0                 8.0     Frequent Customer
2       709                     6.0                 7.0     Frequent Customer
3       764                     7.0                 9.0     Frequent Customer
4       764                     9.0                 9.0     Frequent Customer
5       777                    26.0                11.0      Regular Customer
6       825                    30.0                19.0      Regular Customer
7       910                    30.0                 6.0     Frequent Customer
8      1052                    19.0                 9.5     Frequent Customer
9      1052                    15.0                 9.5     Frequent Customer
10     1379                     7.0                20.0      Regular Customer
11     1480                     7.0                 7.0     Frequent Customer
12     1480                     7.0                 7.0     Frequent Customer
13     1480                    12.0                 7.0     Frequent Customer
14     1494                     2.0                 7.0     Frequent Customer
15     1494                     6.0                 7.0     Frequent Customer
16     1494                    10.0                 7.0     Frequent Customer
17     1540                     7.0                 7.0     Frequent Customer
18     1540                     6.0                 7.0     Frequent Customer
19     1540                    14.0                 7.0     Frequent Customer
20     1540                     8.0                 7.0     Frequent Customer
21     1540                     6.0                 7.0     Frequent Customer
22     1540                    28.0                 7.0     Frequent Customer
23     1540                     7.0                 7.0     Frequent Customer
24     1540                     7.0                 7.0     Frequent Customer
25     1540                     0.0                 7.0     Frequent Customer
26     1540                     7.0                 7.0     Frequent Customer
27     1540                     7.0                 7.0     Frequent Customer
28     1540                     7.0                 7.0     Frequent Customer
29     1540                     7.0                 7.0     Frequent Customer
30     1540                     3.0                 7.0     Frequent Customer
31     1540                    16.0                 7.0     Frequent Customer
32     1540                     5.0                 7.0     Frequent Customer
33     1540                     6.0                 7.0     Frequent Customer
34     1598                     7.0                 6.0     Frequent Customer
35     1598                    21.0                 6.0     Frequent Customer
36     1598                     2.0                 6.0     Frequent Customer
37     1647                     7.0                25.0  Nonfrequent Customer
38     1647                    18.0                25.0  Nonfrequent Customer
39     1647                     8.0                25.0  Nonfrequent Customer
40     1647                    30.0                25.0  Nonfrequent Customer
41     1647                    25.0                25.0  Nonfrequent Customer
42     2833                     7.0                 7.0     Frequent Customer
43     2850                     9.0                 7.0     Frequent Customer
44     2857                    10.0                10.0     Frequent Customer
45     3010                     7.0                 7.0     Frequent Customer
46     3029                     7.0                28.0  Nonfrequent Customer
47     3395                    30.0                18.0      Regular Customer
48     3904                     7.0                15.0      Regular Customer
49     4006                     7.0                 7.0     Frequent Customer
50     4006               

In [93]:
# Step 9 
# Export Dataframe
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_aggregated.pkl'))